docs/_future/custom_functions.md

```diff
@@ -0,0 +1,174 @@
+
+The value for properties in `input_parameters` and `output_parameters` needs to be an object with further sub-properties:
+  * `type`: The type of the input parameter. This can be a [Built-in type](https://api.slack.com/future/types) or a [Custom type](https://api.slack.com/future/types/custom) that you define.
+  * `description`: A string description of the parameter.
+
+If you want to set a property as required, list its name in its respective input or output properties `required` property.
+
+For example, if you have an input parameter named `customer_id` that you want to be required, you can do so like this:
+
+```javascript
+input_parameters: {
+  properties: {
+    customer_id: {
+      type: Schema.types.string,
+      description: "The customer's ID"
+    }
+  },
+  required: ["customer_id"]
+}
+```
+
+Functions can (and generally should) declare inputs and outputs. 
+
+Inputs are declared in the `input_parameters` property, and outputs are declared in the `output_parameters` property. Each can contain either [Built-in types](https://api.slack.com/future/types) or [Custom types](https://api.slack.com/future/types/custom) that you define.
+
+While, strictly speaking, input and output parameters are optional, they are a common and standard way to pass data between functions and nearly any function you write will expect at least one input and pass along an output.
+
+Functions are similar in philosophy to Unix system commands: they should be minimalist, modular, and reusable. Expect the output of one function to eventually become the input of another, with no other frame of reference.
+
+Once your function is defined in its own file in `manifest/functions`, the next steps are to add a function listener, register the listener, and implement the function in a workflow.
+
+---
+
+### Implementing a Function Listener {#implement}
+
+Implement functions in just a few steps:
+
+#### 1. Create the function definition file in the `manifest/functions` directory
+If you haven't done so already, create a file for your function definition to live in and name it something that makes sense for your function. In the [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/blob/future/manifest/function/sample-function.js), we named this file `sample-function.js`
+
+
+#### 2. Add function listener and handler(s)
+Once your function has been defined, you'll need to register a function listener to trigger the function's functionality as well as any other related events.
+
+To do this, create a file in `listeners/functions` directory for your function listener (in this case, the file will be called `hello-world.js`). This file will use the function definition you created earlier and also implement any needed additional logic&mdash;in this case, it'll send a message to greet someone based on what is being passed in as inputs. 
```

Comment:
```suggestion
To do this, create a file in `listeners/functions` directory for your function listener (in this case, the file will be called `hello-world.js`). This file will use the function definition you created earlier and also implement any needed additional logic&mdash;in this case, it'll send a message to greet someone based on what is being passed in as inputs. You can see the full completed `hello-world.js` function listener [here](https://github.com/slack-samples/bolt-js-starter-template/blob/future/listeners/functions/hello-world.js).
```

It looks like we have a link to the full `hello-world.js` file below on line 162. Since it's a bit lengthy, what if we removed the code snippet below and updated this paragraph with this to link out to the snippet?